### DISCUSSION TITLE
Generating gerbers possible?

### DISCUSSION THREAD

**Discussion by saardrimer** [Argmaster/pygerber#288] [General]
Well done with this library; it was nice to have discovered it.

It wasn't immediately clear to me if one would be able to *generate* Gerbers with the library...? If it's not possible now, is it somewhere in the roadmap? :)

Thanks!

**Comment by Argmaster** [comment]
Hi, well, depends what kind of generation you have on your mind, could you elaborate on that?
For now PyGerber can create PNG/SVG images and that's what it's public API can do.
Only option would be to manually construct AST and serialize it, but there is no API for that, you would have to tinker with PyGerber internals.
Coming in 3.0.0 (I'm workig on it) formatter will be added and hence AST to Gerber serialization, which is kind of code generation + AST construction should be easier.
At some point it's possible that ability translate other vector image formats will be introduced (no specific timeline)
At some point API for walking and modifying Gerber files will be introduced (no specific timeline)
At some point something similar to gerber-writer library may be incorporated into PyGerber (I talked with the author of the library and we consider this a viable option, but there is no timeline)

PyGerber is NOT and never will be a PCB editor with GUI.

🙂

**Reply by saardrimer** [reply]
Sorry for not being clearer in my original message. What I need is what (I think) pygerber is doing now, but in reverse :) So I'd give it a dict (or some other structured data) that 'describes' a Gerber (shapes, apertures, flashes, comments, document properties, etc.) and it will yield a well-formatted Gerber string.

Years ago I created something similar (SVG -> Gerber and Gerber->SVG) that was specific to my application; it worked, but was rather primitive and flaky:

  https://github.com/boldport/pcbmode/blob/master/pcbmode/utils/gerber.py

For my next project I don't need the SVG->Gerber, just what I described above; since you've basically done everything so nicely and up-to-the-latest-spec, I was exploring if it was an option for me.

**Reply by Argmaster** [reply]
Sure, please have a look at API I mentioned in previous comment. If it won't do the trick for you, we can try to agree on what the format of data would have to look like for PyGerber to yield what you need. I'm pretty sure you should look into 3.0.0-alpha capabilities though, since in 2.4.x formatting wasn't particularly well implemented and implementing it would be painful. On the other hand in 3.0.0-alpha there is already Formatter class implemented which not only dumps valid Geber code but also gives you various options to define what "pretty" means for you. 

To install development version directly from repository you could use syntax deacribed here:

https://stackoverflow.com/questions/20101834/pip-install-from-git-repo-branch

I'm moving surprisingly fast with the implementation, alpha version may even be deployed within 2 weeks, depending on how many surprises I encounder down the road 🙃

**Reply by saardrimer** [reply]
Great! I'm probably a month away from having the rest of the work -- that will be open source -- ready for the Gerber generation... so no real urgency, particularly since you seem to have all the infrastructure ready :)

**Comment by Argmaster** [comment]
Hi @saardrimer,

As part of PyGerber 3.0.0-alpha development I started to experiment with design of API for generating Gerber code. If you wish to share your thoughts on that, please have a look: 
- [source](https://github.com/Argmaster/pygerber/blob/feature/pygerber-3-compiler/src/pygerber/gerberx3/ast/builder.py)
- [use example in tests](https://github.com/Argmaster/pygerber/blob/feature/pygerber-3-compiler/test/gerberx3/test_ast/test_builder.py)

API is still work in progress, but general sketch of approach is already there.

**Comment by Argmaster** [comment]
Hi, PyGerber 3.0.0a2 is available on PyPI.

```
pip install pygerber==3.0.0a2
```

APIs I mentioned before are available in `pygerber.builder.gerber` module. There is [documentation with examples available online for that](https://argmaster.github.io/pygerber/latest/80_code_generation/10_gerber.html). API is not exhaustive yet, there are at least arc traces and regions (copper pours) missing, I will look into adding them in 3.0.0a3. You will see the progress in release notes. I will close this discussion for now, feel free to open new one in case you would need help.
